edge-modules/edge-dashboard/src/modules.rs

```diff
@@ -0,0 +1,235 @@
+                    .map_err(ErrorInternalServerError)
+                    .and_then(|url| ModuleClient::new(&url).map_err(ErrorInternalServerError)) // can't connect to the endpoint
+                    .map(move |mod_client| {
+                        mod_client
+                            .logs(module_id, &LogOptions::new())
+                            .map_err(ErrorInternalServerError)
+                            .and_then(|data| {
+                                data.map_err(ErrorInternalServerError)
+                                    .fold(BytesMut::new(), |mut acc, chunk| {
+                                        let stream = chunk.as_ref();
+                                        if stream.len() >= 8 {
+                                            let (_, right) = stream.split_at(8);
+                                            acc.extend_from_slice(right);
+                                        }
+                                        Ok::<_, ActixError>(acc)
+                                    })
+                                    .and_then(|body| {
+                                        let mut clone = body.freeze().to_vec().clone();
+                                        clone.retain(|&byte| (byte as char).is_ascii());
+                                        if let Ok(content) = String::from_utf8(clone) {
+                                            HttpResponse::Ok().body(content)
+                                        } else {
+                                            HttpResponse::ServiceUnavailable()
+                                                .body("Logs unable to be displayed")
+                                        }
+                                    })
+                            })
+                    })
+                    .into_future()
+                    .flatten(),
+            )
+        })
+        .unwrap_or_else(|err| {
+            Either::B(ok(HttpResponse::ServiceUnavailable()
+                .content_type("text/plain")
+                .body(format!("{:?}", err))))
+        });
+
+    Box::new(response)
+}
+
+pub fn get_modules(
+    context: web::Data<Arc<Context>>,
+    info: web::Query<AuthRequest>,
+) -> Box<dyn Future<Item = HttpResponse, Error = ActixError>> {
+    return_modules(context, &info.api_version, module_response)
+}
+
+fn module_response(mods: Vec<Module>) -> HttpResponse {
+    HttpResponse::Ok()
+        .content_type("text/json")
+        .body(format!("{:?}", mods))
+}
+
+pub fn get_health(
+    context: web::Data<Arc<Context>>,
+    info: web::Query<AuthRequest>,
+) -> Box<dyn Future<Item = HttpResponse, Error = ActixError>> {
+    return_modules(context, &info.api_version, health_response)
+}
+
+fn health_response(mods: Vec<Module>) -> HttpResponse {
+    let mut device_status = Status::new();
+    let mut edge_agent = false;
+    let mut edge_hub = false;
+    let mut other = true;
+
+    for module in mods.iter() {
```

Comment:
I think this loop does too many things. What about making it more single responsible?

```
fn health_response(mods: Vec<Module>) -> HttpResponse {
    let edgeagent = mods
        .iter()
        .any(|module| module.name() == "edgeAgent" && module.status() == "running");

    let edgehub = mods
        .iter()
        .any(|module| module.name() == "edgeHub" && module.status() == "running");

    let other = mods.iter().any(|module| module.status() != "running");

    let mut device_status = Status::new();
    device_status.set_iotedged();
    device_status.set_edge_agent(edgeagent);
    device_status.set_edge_hub(edgehub);
    device_status.set_other_modules(edgeagent && edgehub && other);

    let health = device_status.return_health();
    HttpResponse::Ok().body(format!(
        "Device health: {:?}\nDevice details: {:?}",
        health, device_status
    ))
}
```
some benefits:
- you don't need mutate variables anymore
- each statement is declarative and basically states what should be done not how it should be done
- a little bit shorter